db/db_impl/db_impl_write.cc

```diff
@@ -1959,7 +1975,70 @@ void DBImpl::NotifyOnMemTableSealed(ColumnFamilyData* /*cfd*/,
 // REQUIRES: this thread is currently at the front of the 2nd writer queue if
 // two_write_queues_ is true (This is to simplify the reasoning.)
 Status DBImpl::SwitchMemtable(ColumnFamilyData* cfd, WriteContext* context,
-                              std::string replication_sequence) {
+                              const MemTableSwitchRecord& mem_switch_record) {
+  mutex_.AssertHeld();
+  MemTable* new_mem = nullptr;
+
+  // Recoverable state is persisted in WAL. After memtable switch, WAL might
+  // be deleted, so we write the state to memtable to be persisted as well.
+  Status s = WriteRecoverableState();
+  if (!s.ok()) {
+    return s;
+  }
+
+  // Attempt to switch to a new memtable and trigger flush of old.
+  assert(versions_->prev_log_number() == 0);
+
+  const MutableCFOptions mutable_cf_options = *cfd->GetLatestMutableCFOptions();
+
+  // Set memtable_info for memtable sealed callback
+#ifndef ROCKSDB_LITE
+  MemTableInfo memtable_info;
+  memtable_info.cf_name = cfd->GetName();
+  memtable_info.first_seqno = cfd->mem()->GetFirstSequenceNumber();
+  memtable_info.earliest_seqno = cfd->mem()->GetEarliestSequenceNumber();
+  memtable_info.num_entries = cfd->mem()->num_entries();
+  memtable_info.num_deletes = cfd->mem()->num_deletes();
+#endif  // ROCKSDB_LITE
+  int num_imm_unflushed = cfd->imm()->NumNotFlushed();
+  SequenceNumber seq = versions_->LastSequence();
+  new_mem = cfd->ConstructNewMemtable(mutable_cf_options, seq);
+  assert(new_mem != nullptr);
+  context->superversion_context.NewSuperVersion();
+  ROCKS_LOG_INFO(immutable_db_options_.info_log,
+                 "[%s] New memtable created with log file: #%" PRIu64
+                 ". Immutable memtables: %d.\n",
+                 cfd->GetName().c_str(), mem_switch_record.next_log_num, num_imm_unflushed);
+  log_write_mutex_.Lock();
+  logfile_number_ = mem_switch_record.next_log_num;
+  // alive_log_files_ and logs_ size should be 1, so updating front is exactly the same as
+  // updating back
+  alive_log_files_.front().number = mem_switch_record.next_log_num;
```

Comment:
There is one minor concern here. As the next_log_num keeps increasing, we always create a new wal file whenever a server is restarted, these empty WAL files will never get recycled. Probably need to add some functionality to cleanup these files (for example, whenever server is restarted)

Reply:
I'm pretty sure this is already the case. Every time we restart the database the next_log_num is set to next_file_number and a new one is created. RocksDB should clean them up in `PurgeObsoleteFiles()` I believe.